tests/test_motd/test_transformers.py

```diff
@@ -0,0 +1,145 @@
+from __future__ import annotations
+
+from collections.abc import Callable
+
+import pytest
+
+from mcstatus.motd import Motd, RawResponseDescription
+from mcstatus.motd.components import MinecraftColor
+from mcstatus.motd.transformers import AnsiTransformer, HtmlTransformer, MinecraftTransformer, PlainTransformer
+
+
+class TestSomethingToSomethingTables:
```

Comment:
This insures that we "duplicate" code properly, we indeed should. But after some change in transformation logic, I often just copy and paste actual output into expected, as the output is too unreadable for human.

Reply:
I mean, what I'd suggest is tests like this:

```python
from mcstatus.motd.components import MinecraftColor
from mcstatus.motd.transformers import AnsiTransformer


def ansi_from_rgb(r: int, g: int, b: int) -> str:
    return f"\033[38;2;{r};{g};{b}m"


tr = AnsiTransformer()
res = tr.transform([MinecraftColor.MINECOIN_GOLD])
res = res.removesuffix("\033[0m").removesuffix("\033[0m").removeprefix("\033[0m")

assert res == ansi_from_rgb(221, 214, 5)
```

You don't need to test multiple elements at once, and you could just make a parametrized tests over single individual colors, and their appropriate RGB colors. This is much better than explicitly testing constants, and i'ts quite readable without really any pasting of the outputs that we're currently getting. 

The only annoyance here is that we have to remove the reset ANSI sequences (there's 2 at beginning and 1 at the end), which is something that might potentially change later (as we don't really need 2 reset sequences at the start, but it's also not an issue), and this change shouldn't fail the test.

Perhaps even just testing for the color being present in the output string would be fine. Either way, it's much better than testing constants.

Reply:
> You don't need to test multiple elements at once, and you instead just make parametrized tests over single individual colors, and their appropriate RGB colors.

While it's much more readable, it's more unstable. If someone changes constants, and tests will automatically adapt to this change, the only way to test the change - manual one. I think, it's much better to keep hard-coded manually tested examples, that will work. If someone changes it - reviewer should check the output manually. (I could add comment about this)

> The only annoyance here is that we have to remove the reset ANSI sequences (there's 2 at beginning and 1 at the end), which is something that might potentially change later (as we don't really need 2 reset sequences at the start, one is sufficient, but it's also not an issue), and this change shouldn't fail the test.

Those sequences should be tested too. We add `Formatting.RESET` after every color in `Motd.parse`, and we add one more reset after every color while transforming. This is because, if someone uses `.simplify` and then `.ansi`, it will break the output to stop resetting formatting (e.g. bold) after every color (as it does Minecraft). While it's quite annoying to have these two resets in tests, they are matter, and we shouldn't touch simplifies code here (to remove one of them).

> It might actually be worth it to check how something like colorama is doing tests

Colorama just [tests constants](https://github.com/tartley/colorama/blob/21c4b94fe21ce29c85c896ace828da24b7527641/colorama/tests/ansi_test.py#L26-L44), too.

> perhaps there's even a way to get the actual color from how this would be rendered in the end. Though if it's too complex, it's not worth it and the above would be enough, we're not a coloring library after all.

I think there is, like, some manual testing
```py
print(result)
assert input("Please check that colors were rendered correctly. Answer 'y'.") == "y"
```
But automating it would be too complex for our purporses.

Reply:
> While it's much more readable, it's more unstable. If someone changes constants, and tests will automatically adapt to this change, the only way to test the change - manual one. I think, it's much better to keep hard-coded manually tested examples, that will work. If someone changes it - reviewer should check the output manually. (I could add comment about this)

I disagree, we can catch a change in a color definition during the review process, we don't need the tests to do it for us. I'd say it's reasonable to assume that if someone would be changing the constant, they're doing so for a reason, i.e. we accidentally used the wrong rgb somewhere. I don't think they should now have to also change the tests.

It's also just code duplication, we have the constants already defined, so why make a test where all you do is literally copy over the values of those constants, and check that they really are set to those values. That just seems incredibly redundant, and all it does is gives more work to someone having a good reason to change a constant.

> I think there is, like, some manual testing
> ```py
> print(result)
> assert input("Please check that colors were rendered correctly. Answer 'y'.") == "y"
> ```
>
> But automating it would be too complex for our purporses.

Yeah let's not do that then.

I still fundamentally disagree with the idea of testing constants, all if is is an additional place where this constant is defined. It's a constant for a reason, there's no reason to compare it to itself for validity. There is however a reason to test that the functions use these constants properly, and so a test like what I suggested would make much more sense IMO.

Reply:
Deleted these tests [here](https://github.com/py-mine/mcstatus/compare/04dfac0d8e018ee365c206f6bc9183ed07570a23..ab53f18be1319577afcde0a1d44e29b6704e0b03).